openwisp_controller/connection/tests/test_admin.py

```diff
@@ -0,0 +1,118 @@
+import json
+
+from django.test import TestCase
+from django.urls import reverse
+
+from ...config.models import Template
+from ...config.tests.test_admin import TestAdmin as TestConfigAdmin
+from ...tests.utils import TestAdminMixin
+from .. import settings as app_settings
+from ..models import Credentials, DeviceConnection, DeviceIp
+from .base import CreateConnectionsMixin, SshServerMixin
+
+
+class TestAdmin(TestAdminMixin, CreateConnectionsMixin,
+                SshServerMixin, TestCase):
+    template_model = Template
+    credentials_model = Credentials
+    deviceip_model = DeviceIp
+    connection_model = DeviceConnection
+    operator_permission_filters = [
+        {'codename__endswith': 'config'},
+        {'codename__endswith': 'device'},
+        {'codename__endswith': 'template'},
+        {'codename__endswith': 'connection'},
+        {'codename__endswith': 'credentials'},
+        {'codename__endswith': 'device_connection'},
+        {'codename__endswith': 'device_ip'},
+    ]
+    _device_params = TestConfigAdmin._device_params.copy()
+
+    def _get_device_params(self, org):
+        p = self._device_params.copy()
+        p['organization'] = org.pk
+        return p
+
+    def test_device_config_update(self):
+        org1 = self._create_org(name='org1')
+        cred = self._create_credentials_with_key(organization=org1, port=self.ssh_server.port)
+        device = self._create_device(organization=org1)
+        update_strategy = app_settings.UPDATE_STRATEGIES[0][0]
+        c = self._create_config(device=device)
+        dc = self._create_device_connection(device=device,
+                                            credentials=cred,
+                                            update_strategy=update_strategy)
+        self._create_device_ip(device=device,
+                               address=self.ssh_server.host)
+        uid = self.ssh_server.users
+        config = json.dumps({
+            'interfaces': [
+                {
+                    'name': 'eth0',
+                    'type': 'ethernet',
+                    'addresses': [
+                        {
+                            'family': 'ipv4',
+                            'proto': 'dhcp'
+                        }
+                    ]
+                }
+            ]
+        })
+        c.config = config
+        dc.update_config()
+        self.ssh_server.client(*uid).exec_command('/etc/init.d/openwisp_config restart')
+        self.assertEqual(device.status, 'applied')
```

Comment:
these 3 previous lines have several issues:

`dc.update_config()` is called automatically by a signal receiver when a configuration is saved, but you're not saving the configuration, so the `status` will remain `applied` and will not switch to `changed`.

What does the line `self.ssh_server.client(*uid).exec_command('/etc/init.d/openwisp_config restart')` do?
It seems like it executes that command on the mock ssh server, which is not what we need to do, because `update_config()` will do that (which will be executed automatically after saving), we need to check that the command has been called in some way, but we can do this later.

So to recap you need to do something like:

```python

c.full_clean()
c.save()
device.refresh_from_db()
self.assertEqual(device.status, 'applied')
```

Reply:
Thanks for reviewing. will correct it ASAP